pkg/logql/log/parser.go

```diff
@@ -38,154 +37,156 @@ type JSONParser struct {
 		return line, true
 	}
+
 	return line, true
 }
 
-func (j *JSONParser) readObject(it *jsoniter.Iterator) error {
-	// we only care about object and values.
-	if nextType := it.WhatIsNext(); nextType != jsoniter.ObjectValue {
-		return errUnexpectedJSONObject
-	}
-	_ = it.ReadMapCB(j.parseMap(""))
-	if it.Error != nil && it.Error != io.EOF {
-		return it.Error
+func (j *JSONParser) handleJson(preKey string, line []byte) error {
+	err := jsonparser.ObjectEach(line, func(key []byte, val []byte, valueType jsonparser.ValueType, offset int) error {
+		jsonKey := preKey + sanitizeLabelKey(string(key), true)
+		if valueType == jsonparser.Array {
+			return nil
+		}
+		if valueType == jsonparser.Object {
+			return j.handleJson(jsonKey+"_", val)
+		}
+		jsonErr := j.setJSONVal(jsonKey, val)
+		if jsonErr != nil {
+			return jsonErr
+		}
+		return nil
+	})
+	if err != nil {
+		return err
 	}
 	return nil
 }
 
-func (j *JSONParser) parseMap(prefix string) func(iter *jsoniter.Iterator, field string) bool {
-	return func(iter *jsoniter.Iterator, field string) bool {
-		switch iter.WhatIsNext() {
-		// are we looking at a value that needs to be added ?
-		case jsoniter.StringValue, jsoniter.NumberValue, jsoniter.BoolValue:
-			j.parseLabelValue(iter, prefix, field)
-		// Or another new object based on a prefix.
-		case jsoniter.ObjectValue:
-			if key, ok := j.nextKeyPrefix(prefix, field); ok {
-				return iter.ReadMapCB(j.parseMap(key))
-			}
-			// If this keys is not expected we skip the object
-			iter.Skip()
-		default:
-			iter.Skip()
-		}
-		return true
+func (j *JSONParser) parseJSONKeyVal(line []byte, requiredLabels []string) error {
+	//check for "jsonParser-not json line" benchmark
+	if strings.Index(string(line), "{") != 0 {
+		return errors.New("illegal format")
 	}
-}
-
-func (j *JSONParser) nextKeyPrefix(prefix, field string) (string, bool) {
-	// first time we add return the field as prefix.
-	if len(prefix) == 0 {
+	for _, field := range requiredLabels {
 		field = sanitizeLabelKey(field, true)
-		if j.lbs.ParserLabelHints().ShouldExtractPrefix(field) {
-			return field, true
-		}
-		return "", false
-	}
-	// otherwise we build the prefix and check using the buffer
-	j.buf = j.buf[:0]
-	j.buf = append(j.buf, prefix...)
-	j.buf = append(j.buf, byte(jsonSpacer))
-	j.buf = append(j.buf, sanitizeLabelKey(field, false)...)
-	// if matches keep going
-	if j.lbs.ParserLabelHints().ShouldExtractPrefix(unsafeGetString(j.buf)) {
-		return string(j.buf), true
-	}
-	return "", false
-}
-
-func (j *JSONParser) parseLabelValue(iter *jsoniter.Iterator, prefix, field string) {
-	// the first time we use the field as label key.
-	if len(prefix) == 0 {
-		key, ok := j.keys.Get(unsafeGetBytes(field), func() (string, bool) {
-			field = sanitizeLabelKey(field, true)
-			if !j.lbs.ParserLabelHints().ShouldExtract(field) {
-				return "", false
+		keys := make([]string, 0)
+		keys = append(keys, field)
+
+		cacheKeys, ok := j.jsonKeyCache[field]
```

Comment:
Hm, this seems to be an orthogonal improvement to the parser. Could you factor out the cache in another PR? The benchmarks do not seem to be a fair comparison between the different parser implementations. How much improvements does the cache yield?

Reply:
buger/jsonparser with cache : 1172 ns/op
buger/jsonparser without cache: 2223 ns/op
json-iterator:  2913 ns/op

Reply:
cache is very effective for this json library, especially when the json key is deep.

but two problems that the test could not pass. give me some hints.
FAIL: Test_jsonParser_Parse/escaped
FAIL: Test_jsonParser_Parse/utf8_error_rune

Reply:
Thanks for the patience!